clippy_lints/src/direct_recursion.rs

```diff
@@ -0,0 +1,245 @@
+        self.expr_check_blocker.is_some()
+    }
+
+    /// Tries opening the lock using `expr` as the key.
+    fn try_unlocking_with(&mut self, expr: &Expr<'_>) {
+        if let Some(key) = self.expr_check_blocker
+            && key == expr.hir_id
+        {
+            self.expr_check_blocker = None;
+        }
+    }
+}
+
+impl_lint_pass!(DirectRecursion => [DIRECT_RECURSION]);
+
+impl<'tcx> LateLintPass<'tcx> for DirectRecursion {
+    /// Whenever we enter a Body, we push its owner's `DefId` into the stack
+    fn check_body(&mut self, cx: &LateContext<'tcx>, body: &Body<'_>) {
+        self.fn_id_stack.push(cx.tcx.hir_body_owner_def_id(body.id()));
+    }
+
+    /// We then revert this when we exit said `Body`
+    fn check_body_post(&mut self, _: &LateContext<'tcx>, _: &Body<'_>) {
+        _ = self.fn_id_stack.pop();
+    }
+
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'tcx>) {
+        // We use this inner lock so that we avoid recursing if we've
+        // already linted an expression that contains the expression
+        // we're now visiting.
+        // This lock is closed whenever we emit a lint, and it's opened
+        // after we exit the node that closed it (see `check_expr_post`)
+        if self.is_blocked() {
+            return;
+        }
+
+        // Before running the lint, we look up the attributes of this Expr.
+        // If it has been marked with `clippy::allowed_recursion`, then
+        // we ignore it, as well as everything inside it; someone has
+        // decided that the recursive calls within it are fine.
+        let attrs = cx.tcx.hir_attrs(expr.hir_id);
+        if get_attr(cx.sess(), attrs, sym::allowed_recursion).next().is_some() {
+            self.block_with_expr(expr);
+            return;
+        }
+
+        match expr.kind {
+            ExprKind::MethodCall(_, _, _, _) => {
+                // Uniquely identifying the `DefId` of method calls requires doing type checking.
+                // `cx` takes care of that, and then we can ask for the `type_dependent_def_id`
+                // of the `expr` whose kind is `ExprKind::MethodCall`.
+                let typeck = cx.typeck_results();
+                if let Some(type_resolved_def_id) = typeck.type_dependent_def_id(expr.hir_id) {
+                    for stored_fn_id in &self.fn_id_stack {
+                        if stored_fn_id.to_def_id() == type_resolved_def_id {
+                            span_lint(cx, DIRECT_RECURSION, expr.span, "this method contains a call to itself");
+                            self.block_with_expr(expr);
+                            return;
+                        }
+                    }
+                }
+            },
+            ExprKind::Call(path_expr, _) => {
+                // This should almost always be true, as far as I'm aware.
+                // `ExprKind::Call` values are supposed to contain an `Expr` of type `ExprKind::Path`
+                // inside of them.
```

Comment:
Why do you think that? Couldn't that be a field access, a function call, or any expression that returns something implementing `FnOnce`? This comment should probably be removed.